src/QuadForm/Quad/ZGenus.jl

```diff
@@ -0,0 +1,2001 @@
+  p = G._prime
+  s = div(n + 1, 2)
+  std = 2*prod(fmpq[1 - QQ(p)^(-2*k) for k in 1:s-1])
+  if n % 2 == 0
+    D = ZZ(-1)^s * ZZ(determinant(G))
+    epsilon = _kronecker_symbol(4*D, p)
+    std *= (1 - epsilon*QQ(p)^(-s))
+  end
+  return QQ(1) // std
+end
+
+@doc Markdown.doc"""
+    _species_list(G::ZpGenus)
+
+Return the species list.
+See Table 1 in [CS1988]_.
+"""
+function _species_list(G::ZpGenus)
+  p = prime(G)
+  species_list = []
+  sym = G._symbol
+  if p != 2
+    for k in 1:length(sym)
+      n = ZZ(sym[k][2])
+      d = sym[k][3]
+      if n % 2 == 0 && d != _kronecker_symbol(-1, p)^(div(n, 2))
+        species = -n
+      else
+        species = n
+      end
+      push!(species_list, species)
+    end
+    return species_list
+  end
+  #  p == 2
+  # create a dense list of symbols
+  symbols = Vector{Int}[]
+  s = 1
+  for k in 0:sym[end][1]
+    if sym[s][1] == k
+      push!(symbols, sym[s])
+      s +=1
+    else
+      push!(symbols,[k, 0, 1, 0, 0])
+    end
+  end
+  # avoid a case distinction
+  sym = [[-2, 0, 1, 0, 0],[-1, 0, 1, 0, 0]]
+  append!(sym, symbols)
+  push!(sym, [sym[end-1][1] + 1, 0, 1, 0, 0])
+  push!(sym, [sym[end-1][1] + 2, 0, 1, 0, 0])
+  for k in 2:length(sym)-1
+    free = true
+    if sym[k-1][4]==1 || sym[k+1][4]==1
+      free = false
+    end
+    n = sym[k][2]
+    o = sym[k][5]
+    if _kronecker_symbol(sym[k][3], 2) == -1
+      o = mod(o + 4, 8)
+    end
+    if sym[k][4] == 0 || n % 2 == 1
+      t = div(n, 2)
+    else
+      t = div(n, 2) - 1
+    end
+    if free && (o == 0 || o == 1 || o == 7)
+      species = 2*t
+    elseif free && (o == 3 || o == 5 || o == 4)
+      species = -2*t
+    else
+      species = 2*t + 1
+    end
+    push!(species_list, species)
+  end
+  return species_list
+end
+
+
+
+
+@doc Markdown.doc"""
+    _gamma_exact(n)
+
+Evaluates the exact value of the `\Gamma^2` function at an integer or
+half-integer argument. Ignoring factors of pi
+"""
+function _gamma_exact(n)
+  n = QQ(n)
+  if denominator(n) == 1
+    @req (n > 0) "not in domain"
+    return factorial(ZZ(n) - 1)
+  end
+  @req (denominator(n) == 2) "n must be in (1/2)ZZ"
+  a = QQ(1)
+  while n != 1//2
+    if n < 0
+      a //= n
+      n += 1
+    elseif n > 0
+      n += -1
+      a *= n
+    end
+  end
+  return a
+end
+
+@doc Markdown.doc"""
+    _zeta_exact(n)
+
+Returns the exact value of the Riemann Zeta function
+ignoring factors of pi.
+
+The argument must be a critical value, namely either positive even
+or negative odd.
+
+See for example [Iwa1972]_, p13, Special value of `\zeta(2k)`
+REFERENCES:
+
+- [Iwa1972]_
+- [IR1990]_
+- [Was1997]_
+"""
+function _zeta_exact(n)
+  if n < 0
+    return bernoulli(1-n)//(n-1)
+  elseif n > 1
+    if (n % 2 == 0)
+      return QQ(-1)^(div(n, 2) + 1) * QQ(2)^(n-1) * bernoulli(n) // factorial(ZZ(n))
+    else
+      raise(error("n must be a critical value (i.e. even > 0 or odd < 0)"))
+    end
+  elseif n == 1
+    raise(error("Here is a pole"))
```

Comment:
```suggestion
    error("Here is a pole")
```
There is no raise function.